torchtune/models/llava_next/image_processor.py

```diff
@@ -0,0 +1,344 @@
+                "resample must be of type torchvision.transforms.InterpolationMode or ['bilinear', 'bicubic]."
+            )
+
+        self.center_crop = v2.CenterCrop(size=patch_size)
+
+        self.patch_size = patch_size
+
+    @staticmethod
+    def _get_new_size_without_distortion(
+        image_size: Tuple[int, int],
+        target_resolution: List[int],
+    ) -> Tuple[int, int]:
+
+        """
+        Determines the maximum dimensions to which an image can be resized without distorting its
+        aspect ratio, based on the target resolution.
+
+        Args:
+            image_size (Tuple[int, int]): The original dimensions of the image (height, width).
+            target_resolution (List[int]): The desired resolution to fit the image into (height, width).
+        Returns:
+            Tuple[int, int]: The optimal dimensions (height, width) to which the image should be resized.
+        Example:
+            >>> _get_new_size_without_distortion([200, 300], target_resolution = [450, 200])
+            (134, 200)
+            >>> _get_new_size_without_distortion([800, 600], target_resolution = [450, 1300])
+            (450, 338)
+        """
+
+        original_height, original_width = image_size
+        target_height, target_width = target_resolution
+
+        scale_w = target_width / original_width
+        scale_h = target_height / original_height
+
+        if scale_w < scale_h:
+            new_width = target_width
+            new_height = min(math.ceil(original_height * scale_w), target_height)
+        else:
+            new_height = target_height
+            new_width = min(math.ceil(original_width * scale_h), target_width)
+
+        return new_height, new_width
+
+    @staticmethod
+    def _divide_to_patches(image: torch.Tensor, patch_size: int) -> List[torch.Tensor]:
+        """
+        Divides an image into patches. The patches will not have equal size if the
+        aspect ratio of the image is not divisible by the patch size.
+
+        Args:
+            image (torch.Tensor): Image tensor.
+            patch_size (int): Size of each patch.
+        Returns:
+            List[torch.Tensor]: List of image patches.
+        Example:
+            >>> image = torch.rand(3, 200, 300)
+            >>> patches = _divide_to_patches(image, patch_size=50)
+            >>> len(patches)
+            24
+            >>> image = torch.rand(3, 400, 500)
+            >>> patches = _divide_to_patches(image, patch_size=450)
+            >>> len(patches)
+            2
+        """
+
+        channel_size, height, width = image.shape
+
+        patches = []
+        for i in range(0, height, patch_size):
+            for j in range(0, width, patch_size):
+                patch = image[:, i : i + patch_size, j : j + patch_size]
+                patches.append(patch)
+
+        return patches
+
+    @staticmethod
+    def _select_best_resolution(
+        original_size: Tuple[int, int], possible_resolutions: torch.Tensor
```

Comment:
you've made possible_resolutions List[Tuple[int, int]] elsewhere, why tensor here?

Reply:
good catch! There are 3 places where possible_resolutions is an input:

1. ImageProcessor
2. GetImagePatches
3. _select_best_resolution

The user enters this in ImageProcessor, which uses it to initialize GetImagePatches.

in __init__ of GetImagePatches, I make it a tensor. Then, for every call of _select_best_resolution, we can use it as a tensor, since this function is vectorized.

We could ask the user to input it a torch.Tensor in ImageProcessor, but I feel that it would make their lives harder.

Does it make sense?